modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermarkImpl.java

```diff
@@ -0,0 +1,298 @@
+            lastScheduledTaskFuture.cancel(true);
+        }
+
+        IgniteUtils.shutdownAndAwaitTermination(scheduledThreadPool, 10, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public @Nullable HybridTimestamp getLowWatermark() {
+        return lowWatermark;
+    }
+
+    CompletableFuture<Void> updateLowWatermark() {
+        return inBusyLock(busyLock, () -> {
+            HybridTimestamp lowWatermarkCandidate = createNewLowWatermarkCandidate();
+
+            // Wait until all the read-only transactions are finished before the new candidate, since no new RO transactions could be
+            // created, then we can safely promote the candidate as a new low watermark, store it in vault, and we can safely start cleaning
+            // up the stale/junk data in the tables.
+            return txManager.updateLowWatermark(lowWatermarkCandidate)
+                    .thenComposeAsync(unused -> inBusyLock(busyLock, () -> {
+                        vaultManager.put(LOW_WATERMARK_VAULT_KEY, ByteUtils.toBytes(lowWatermarkCandidate));
+
+                        lowWatermark = lowWatermarkCandidate;
+
+                        return notifyListeners(lowWatermarkCandidate);
+                    }), scheduledThreadPool)
+                    .whenComplete((unused, throwable) -> {
+                        if (throwable != null) {
+                            if (!(throwable instanceof NodeStoppingException)) {
+                                LOG.error("Failed to update low watermark, will schedule again: {}", throwable, lowWatermarkCandidate);
+
+                                inBusyLock(busyLock, this::scheduleUpdateLowWatermarkBusy);
+                            }
+                        } else {
+                            LOG.info("Successful low watermark update: {}", lowWatermarkCandidate);
+
+                            scheduleUpdateLowWatermarkBusy();
+                        }
+                    });
+        });
+    }
+
+    @Override
+    public void addUpdateListener(LowWatermarkChangedListener listener) {
+        updateListeners.add(listener);
+    }
+
+    @Override
+    public void removeUpdateListener(LowWatermarkChangedListener listener) {
+        updateListeners.remove(listener);
+    }
+
+    private CompletableFuture<Void> notifyListeners(HybridTimestamp lowWatermark) {
+        if (updateListeners.isEmpty()) {
+            return nullCompletedFuture();
+        }
+
+        lwmLock.lock();
+        try {
+            var res = new ArrayList<CompletableFuture<?>>();
+            for (LowWatermarkChangedListener updateListener : updateListeners) {
+                res.add(updateListener.onLwmChanged(lowWatermark));
+            }
+
+            return CompletableFuture.allOf(res.toArray(CompletableFuture[]::new));
+        } finally {
+            lwmLock.unlock();
+        }
+    }
+
+    @Override
+    public void getLowWatermarkSafe(Consumer<HybridTimestamp> consumer) {
+        lwmLock.lock();
```

Comment:
I think the low watermark may still change while the consumer is running. 
I think we need to take a lock on update the field `lowWatermark`.

Reply:
Done.
I don't like this approach, anyway.